Source/Foundation/bsfUtility/Utility/BsDynLibManager.cpp

```diff
@@ -1,60 +1,44 @@
 //************************************ bs::framework - Copyright 2018 Marko Pintera **************************************//
 //*********** Licensed under the MIT license. See LICENSE.md for full terms. This notice is not to be removed. ***********//
 #include "Utility/BsDynLibManager.h"
-#include "Utility/BsDynLib.h"
 
 namespace bs
 {
-	DynLib* DynLibManager::load(const String& name)
+	DynLib* DynLibManager::load(String filename)
 	{
 		// Add the extension (.dll, .so, ...) if necessary.
-		String filename = name;
-		const UINT32 length = (UINT32)filename.length();
+		// TODO: The string comparison here could be slightly more efficent
+		// Using a templatized string_concat function for the lower_bound call
+		// and/or a custom comparitor that does comparison by parts.
+		const String::size_type length = filename.length();
 		const String extension = String(".") + DynLib::EXTENSION;
-		const UINT32 extLength = (UINT32)extension.length();
+		const String::size_type extLength = extension.length();
 		if (length <= extLength || filename.substr(length - extLength) != extension)
-			filename += extension;
+			filename.append(extension);
 
 		if(DynLib::PREFIX != nullptr)
-			filename = DynLib::PREFIX + filename;
+			filename.insert(0, DynLib::PREFIX);
 
-		auto iterFind = mLoadedLibraries.find(filename);
-		if (iterFind != mLoadedLibraries.end())
+		const auto& iterFind = mLoadedLibraries.lower_bound(filename);
```

Comment:
I don't believe this should be a reference, considering the iterator is returned by value, unless I'm misunderstanding which type this gets deduced to.

Reply:
The reference should decay so that the resulting compiled code is simply storing the value.

The main value of using const-& when using the auto keyword is in it's behavior in the general case of various situations.

When the function returns a value, the const-auto-& does "lifetime extension", so that the temporary that's returned lives until the const-auto-& variable goes out of scope. In effect, after the compiler gets done with it, this really just means that the variable is treated as a value, not a reference, and the "lifetime extension of the temporary" is only meaningful in the land of standarese, not in the land of assembly code.

When the function returns a reference, the const-auto-& captures that reference, and prevents any kind of copy-construction (which can potentially involve an allocation, depending on underlying type, and malloc is expensive and blocking). So the avoidance of the copy-construction is desirable, except in the case where we *want* a copy, so that the original object that the reference points to can be modified.

So perhaps it's a bit of a force of habit on my part, but I almost always prefer to use const auto&, or auto&& when storing the return value of a function, except in cases where I explicitly need a copy, and the reason is specifically because in my normal day-to-day work, the codebase I work with isn't consistent on whether a function returns by value, or by reference, and whether copy-construction is cheap or not. For what I do, it's safer to use the "safe" default of const-auto-& than to try to micro manage what I do with the returned variable.

That being said, yes, in this situation, the & doesn't meaningfully change what the compiler is going to do, as the returned value is, i believe, a value.

Reply:
I wasn't aware of this const-& behavior and assumed it's just going to end up being a plain reference. This is definitely fine then!

Reply:
You can read more about it here, if you sift through the discussion: https://stackoverflow.com/questions/19864176/how-would-auto-extend-the-life-time-of-the-temporary-object